FIX: Deep copy MapType in CEL composition to prevent data race

Abhigyan-Shekhar opened the pull request:
```release-note
Fixed a race condition in the CEL compiler that could occur when initializing composited policies concurrently. 

### Description
Fixes a fatal crash (concurrent map read/write) in `NewCompositedCompilerFromTemplate`.

The `NewCompositedCompilerFromTemplate` function previously performed a shallow copy of `CompositionEnv`, sharing the `MapType` pointer across all compilers. Under high concurrency, this caused a race condition when `FindStructFieldType` (reader) and `AddField` (writer) accessed `MapType.Fields` simultaneously, leading to an APIServer panic.

This change implements a deep copy of the `Fields` map for each composition environment, ensuring thread safety.

### Issue
Fixes #135757

```release-note
Fixed a race condition in CEL admission policy compilation that could cause the kube-apiserver to crash with a "concurrent map read and map write" error under high load.

Abhigyan-Shekhar commented:
/release-note
Fixed a race condition in CEL admission policy compilation that could cause the kube-apiserver to crash with a "concurrent map read and map write" error under high load.

BenTheElder commented:
/cc @lalitc375

lalitc375 commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/composition.go:
```diff
@@ -70,8 +70,14 @@ func NewCompositedCompiler(envSet *environment.EnvSet) (*CompositedCompiler, err
 func NewCompositedCompilerFromTemplate(context *CompositionEnv) *CompositedCompiler {
+	mapType := *context.MapType
+	mapType.Fields = make(map[string]*apiservercel.DeclField, len(context.MapType.Fields))
+	for k, v := range context.MapType.Fields {
```
I don't think this iteration is needed. In all combinations fields are always empty. Add a comment we are making fields empty similar to CompiledVariables.

Abhigyan-Shekhar replied:
I have removed the iteration and initialized Fields as an empty map as suggested.

rpb-ant commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/composition_race_test.go:
```diff
@@ -0,0 +1,62 @@
+limitations under the License.
+*/
+
+package cel
```
Non-blocking: the change you are suggesting is a more direct check of the specific data race that we uncovered. I think there is added value in having a _generic_ "are there any data races" test. It is not 100% guaranteed to uncover data races, but if we introduce any new unintended races that type of test _might_ uncover them.

If k8s does not generally run tests in CI with `-race` then it's probably not worth adding.

Abhigyan-Shekhar commented on staging/src/k8s.io/kubectl/pkg/util/resource/resource.go:
```diff
@@ -33,7 +33,19 @@ import (
 // non-zero quantity.
 func PodRequestsAndLimits(pod *corev1.Pod) (reqs, limits corev1.ResourceList) {
-	return podRequests(pod), podLimits(pod)
+	reqs, limits = podRequests(pod), podLimits(pod)
```
I assigned them to variables so  that I could overwrite specific fields with the Pod-level resources (in the if block below) before returning. This implements the merge logic discussed with Brian to handle cases where users mix Pod limits with Container requests.

rpb-ant commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/condition_test.go:
```diff
@@ -803,7 +803,7 @@ func TestCondition(t *testing.T) {
 			results: []EvaluationResult{
 				{
-					Error: errors.New(fmt.Sprintf("operation cancelled: actual cost limit exceeded")),
+					Error: fmt.Errorf("operation cancelled: actual cost limit exceeded"),
```
seems unrelated; agreed that this is a good change, just unnecessary in this PR

rpb-ant commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/composition.go:
```diff
@@ -70,19 +71,44 @@ func NewCompositedCompiler(envSet *environment.EnvSet) (*CompositedCompiler, err
+	})
+	if err != nil {
+		panic(fmt.Errorf("failed to extend environment for composition: %w", err))
+	}
```
I agree that this error condition seems impossible in the code as it is currently written. I think this would be a bit cleaner if we created a `MustExtend` function (or maybe `mustExtend` since it's probably only used in this package) that makes it more clear that this operation is expected to be infallible.

rpb-ant reviewed (approved):
Verified that this PR also passes my data race test, LGTM, three small suggestions

Abhigyan-Shekhar commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/condition_test.go:
```diff
@@ -1573,6 +1573,7 @@ func (f fakeAuthorizer) Authorize(ctx context.Context, a authorizer.Attributes)
 			panic(fmt.Sprintf("unsupported type: %T", a))
 		}
 
+		// Use reflect.DeepEqual to compare structs containing errors safe
```
I verified locally that the tests still pass without this change, so I have reverted it as requested.

Abhigyan-Shekhar commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/composition.go:
```diff
@@ -70,19 +71,44 @@ func NewCompositedCompiler(envSet *environment.EnvSet) (*CompositedCompiler, err
+			cel.Variable("variables", newMapType.CelType()),
+		},
+		DeclTypes: []*apiservercel.DeclType{
+			&newMapType,
```
Okay , I’ve refactored the CEL composition flow to align with #136004.

lalitc375 commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/composition.go:
```diff
@@ -112,47 +168,57 @@ func (c *CompositedCompiler) CompileMutatingEvaluator(expression ExpressionAcces
+// Deprecated: NewCompositionEnv is deprecated.
+// Use NewCompositedCompilerForTypeChecking instead, which properly extends EnvSet
+// once at compiler creation time.
+func NewCompositionEnv(typeName string, envSet *environment.EnvSet) (*CompositionEnv, error) {
```
We can delete this method now. What is stopping us to delete this method?

lalitc375 commented:
Can you squash commits into very small number of commits? and also mark the comments resolved.

Abhigyan-Shekhar committed:
FIX: Deep copy MapType in CEL composition to prevent data race

This commit fixes a fatal crash (concurrent map read/write) in
NewCompositedCompilerFromTemplate by:

- Refactoring CEL EnvSet composition to eliminate cloning and symbol conflicts
- Adding NewCompositedCompilerForTypeChecking for typechecking
- Removing deprecated CompositionEnv type and functions
- Adding regression test for concurrency race condition
- Using mustExtend helper function for clearer intent

Abhigyan-Shekhar commented:
> Can you squash commits into very small number of commits? and also mark the comments resolved.

@lalitc375 I have squashed all the commits to one and resolved all the comments. Thank you for being patient with me and guiding me throughout

lalitc375 commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/composition.go:
```diff
@@ -112,47 +160,26 @@ func (c *CompositedCompiler) CompileMutatingEvaluator(expression ExpressionAcces
-
-func (c *CompositionEnv) AddField(name string, celType *cel.Type) {
-	c.MapType.Fields[name] = apiservercel.NewDeclField(name, convertCelTypeToDeclType(celType), true, nil, nil)
+	typeName          string
```
I think we can remove this typeName now.

lalitc375 commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/validating/validator.go:
```diff
@@ -45,15 +45,17 @@ type validator struct {
 		auditAnnotationFilter: auditAnnotationFilter,
 		messageFilter:         messageFilter,
 		failPolicy:            failPolicy,
+		err:                   err,
```
We can rename it some thing like compileError and add go docs explaining purpose of this error.

lalitc375 commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/composition.go:
```diff
@@ -42,48 +42,96 @@ type CompositedCompiler struct {
+// It initializes the composition state but leaves the Compilers nil, as they are expected
+// to be replaced by the caller (who is doing type checking).
+func NewCompositedCompilerForTypeChecking(envSet *environment.EnvSet) (*CompositedCompiler, error) {
+	newMapType := apiservercel.NewObjectType(VariablesTypeName, map[string]*apiservercel.DeclField{})
```
VariablesTypeName can be pkg scope.

lalitc375 reviewed (commented):
/lgtm

/assign @jpbetz  - Please provide your review.

seans3 commented:
/kind bug

Abhigyan-Shekhar commented:
/release-note
